cognite_toolkit/_cdf_tk/commands/_download.py

```diff
@@ -0,0 +1,92 @@
+from collections import Counter
+from collections.abc import Iterable
+from functools import partial
+from pathlib import Path
+
+from rich.console import Console
+
+from cognite_toolkit._cdf_tk.storageio import StorageIO
+from cognite_toolkit._cdf_tk.storageio._base import T_CogniteResourceList, T_StorageID, T_WritableCogniteResourceList
+from cognite_toolkit._cdf_tk.utils.file import safe_write, to_directory_compatible, yaml_safe_dump
+from cognite_toolkit._cdf_tk.utils.fileio import Compression, FileWriter
+from cognite_toolkit._cdf_tk.utils.producer_worker import ProducerWorkerExecutor
+from cognite_toolkit._cdf_tk.utils.useful_types import JsonVal
+
+from ._base import ToolkitCommand
+
+
+class DownloadCommand(ToolkitCommand):
+    def download(
+        self,
+        identifiers: Iterable[T_StorageID],
+        io: StorageIO[T_StorageID, T_CogniteResourceList, T_WritableCogniteResourceList],
+        output_dir: Path,
+        verbose: bool,
+        file_format: str,
+        compression: str,
+        limit: int | None = 100_000,
+    ) -> None:
+        """Downloads data from CDF to the specified output directory.
+
+        Args:
+            identifiers: The identifiers of the resources to download.
+            io: The StorageIO instance that defines how to download and process the data.
+            output_dir: The directory where the downloaded files will be saved.
+            verbose: If True, prints detailed information about the download process.
+            file_format: The format of the files to be written (e.g., ".ndjson").
+            compression: The compression method to use for the downloaded files (e.g., "none", "gzip").
+            limit: The maximum number of items to download for each identifier. If None, all items will be downloaded.
+        """
+        target_directory = output_dir / io.folder_name
+        target_directory.mkdir(parents=True, exist_ok=True)
```

Comment:
there seems to be some weakness wrt reusing directory or intermediate results stemming from
- allowing destination to have data
- auto-numerating clashes internally
- not using atomic move from temporary clean download location to target resting path

Which may show up as confusing output when running and aborting or running multiple times into same destination dir, it might be not clear what state destination ends up in: partial files, files from previous runs etc

Not clear enough how critical this is, who's the expected caller for downloads (and which precaution steps or assumptions will it have)

Reply:
You have some good points. I have noted them down in [this task](https://cognitedata.atlassian.net/browse/CDF-25572). 

This is a continuation of the current `cdf dump` task, which have the same behavior, so I prefer we keep it as is for now.